### ISSUE TITLE
new rule: calls or nested comprehensions unnecessarily nested within loop

### ISSUE THREAD

**Issue by oliver-sanders** [adamchainz/flake8-comprehensions#663]
### Description

Proposal for a new rule to help spot inefficient comprehensions.

### The Rule

Calls or nested comprehensions within the loop of a comprehension, which do not use arguments derived from the loop, are an inefficiency.

<table>
<thead>
<tr>
<th>Bad:</th>
<th>Good</th>
</tr>
</thead><tbody>
<tr><td>

```python
z = [
    item
    for item in one()
    if item not in two()
]
```

</td><td>

```python
_two = {*two()}
z = [
    item
    for item in one()
    if item not in _two
]
```

</td></tr><tr><td>

```python
z = [
    item
    for item in one()
    if item not in [chr(x) for x in range(50, 55)]
]
```

</td><td>

```python
_chrs = [chr(x) for x in range(50, 55)]
z = [
    item
    for item in one()
    if item not in _chrs
]
```

</td></tr>

</tbody>
</table>

### Rationale / Example

Say we have a couple of generators like so:

```python
def one():
    print('one')
    yield 'a'
    yield 'b'

def two():
    print('two')
    yield 'a'
    yield 'c'
```

We may be tempted to write a comprehension like this:

```python
z = [
    item
    for item in one()
    if item not in two()
]
```

However, this is bad because `two()` is being called repeatability within the loop:

```console
$ python example.py
one
two
two
```

The equivalent `for` loop would be:

```python
z = []
for item in one():
    if item not in two():  # inefficient call
        z.append(item)
```

When written as a `for` loop, the inefficiency is fairly obvious.

But when written as a comprehension, especially as a nested comprehension, the execution logic is less obvious, especially to those less familiar with comprehensions.

### Potential For False Positives

There is no issue if the call/nested-comprehension uses an argument from its outer loop:

```python
# good
z = [
    item
    for number in range(1, 10)
    if item not in [chr(x * number) for x in range(50, 55)]
]

# good
z = [
    item
    for number in range(1, 1]0)
    if item not in [f'{number}-{x}' for x in range(50, 55)]
]
```

So we would have to be careful to ensure we don't get get any false-positives here.

I think all other cases would likely be examples of B307, e.g:

```python
# bad
z= [
    item
    for item in range(10)
    if not print(item)  # B307 - comprehensions should not have side-effects
]
```

I.e, flagging a different anti-pattern.
